### PR TITLE
Integrate with Prometheus's Service Discovery Manager

### LINKED ISSUE #16: Support for k8s service discovery

### PR DESCRIPTION
**Pull request by jcreixell**
- By using Prometheus's SD manager, we create some coupling between
    Periskop and Prometheus (we were already using the DNS resolver from
    Prometheus anyway).
  - This bring also many dependencies from Prometheus.
  - However, in exchange, we get for free ALL service discovery
    mechanisms supported by prometheus out of the box, with a
    configuration that mirrors the one for Prometheus.
  - This should make it a no brainer for anyone already using Prometheus
    to start using Periskop.
  - I have tested it only with DNS and it works fine.
  - See
    https://prometheus.io/docs/prometheus/latest/configuration/configuration/
    for a list of supported service discovery mechanisms and
    configuration formats

Fixes #16 

Maybe @beorn7 has some opinions on this one 🔆

### CODE REVIEW

**Comment by beorn7** [comment]
SD is quite a loaded topic in Prometheus itself.

I guess Periscop should just use it in the same way as [Loki](https://github.com/grafana/loki). Both projects have the same use case: rip out SD from Prometheus and use it in their own way.

**Comment by jcreixell** [comment]
@beorn7 thanks, did not know about Loki. Good to know we are not the only ones doing this :)

File: .cirrus.yml [hunk 1]
```diff
@@ -13,7 +13,7 @@ task:
   name: Lint Api
   container:
     image: golangci/golangci-lint:latest
-  lint_script: golangci-lint run -v --out-format json > lint-report.json
+  lint_script: golangci-lint run --timeout=10m -v --out-format json > lint-report.json
```
Comment (marctc) [inline .cirrus.yml:16]: Good catch!

**Review by marctc — APPROVED** [review]
looks good! Maybe is a good PR to start adding tests in this package :smile:

File: .cirrus.yml (hunk 1 above)
Comment (jcreixell) [reply .cirrus.yml:16]: actually, I don't know why it got so slow TBH

**Comment by jcreixell** [comment]
@beorn7 btw wondering if the SD module from prometheus could be extracted into a separate library? it does not seem to have much prometheus-specific, and seems to be useful for other projects. Anyway, just an idea :)

**Comment by beorn7** [comment]
There have been a lot of discussions about extracting parts of Prometheus into libraries for the benefit of the wider ecosystem (not only Periskop, but Loki, Cortex, Thanos and probably many more use parts of the Prometheus code base). TSDB was the one part where this was actually done. With the result of merging it back in after a while because it turned out to be tedious four Prometheus itself. Now we have the problem that the semantic versioning of Prometheus is meant to be for the resulting binary, not for the interface of the various packages. Therefore, it's at odds with the whole idea of utilizing semantic versioning for Go modules. The (IMHO) most obvious way forward would be to turn the prometheus/prometheus repo into one with multiple Go modules and version them independently from the overall Prometheus server versioning. However, multi-module repos are apparently very tedious, and the whole thing would mean a lot of overhead and friction for developing Prometheus itself. There are currently no plans of going into that direction. Instead, there is an experiment trying to solve the problem in another direction: [Modularise](https://docs.google.com/document/d/1g7wIlxn9JBJkc-2lHAb2MfCYLltWZKyNlIO6SBW1x-8/edit#). I vaguely remember that @Helcaraxan had already set up some repos auto-extracted from Prometheus, but now I cannot find them anymore. Not sure what the current state is.

**Comment by Helcaraxan** [comment]
There are two modules available at the moment, one for TSDB and another for `promql`:
- [`github.com/modularise/prometheus-tsdb`](https://github.com/modularise/prometheus-tsdb)
- [`github.com/modularise/prometheus-promql`](http://github.com/modularise/prometheus-promql)

Other modules can be added very easily. I just need to create the repo and add a few lines of config. Let me know if that is required.

**Comment by Helcaraxan** [comment]
Based on the actual requirement of this PR I've just created the [`github.com/modularise/prometheus-discovery`](https://github.com/modularise/prometheus-discovery) repo / module as well.

It will be automatically synced every day with the latest content from `prometheus/prometheus`.

This will help decrease the impact of this PR on the dependency tree by a ~20-30 modules out of the ~250 that the importing the whole Prometheus project would bring. That means there's still a significant impact on your dependency tree I'm afraid but that is due to the structure of dependencies of the prometheus project itself.

It will however limit the number and size of packages that you expose to the soundcloud/periskop project via its direct dependencies, preventing developers from hijacking other pieces of the prometheus project without requiring a modification to your `go.mod`.

**Comment by jcreixell** [comment]
That's awesome @Helcaraxan! I will give it a try, much appreciated!

**Comment by Helcaraxan** [comment]
Happy to see that the modularised version of the `discovery` functionalities helps. 🙂

**Pull request merged**